internal/telemetry/amplitude.go

```diff
@@ -0,0 +1,128 @@
+package telemetry
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+)
+
+const (
+	amplitudeURL           = "https://api2.amplitude.com/2/httpapi"
+	eventChannelBufferSize = 8
+	eventBufferSize        = 32
+)
+
+var eventSendInterval = time.Second * 15
+
+// AmplitudeClient was inlfuenced by https://github.com/savaki/amplitude-go
+// but uses the version 2 HTTP API.
+type AmplitudeClient struct {
+	apiKey string
+	events chan Event
+	flush  chan chan struct{}
+	client *http.Client
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func NewAmplitudeClient(ctx context.Context, client *http.Client, apiKey string) *AmplitudeClient {
+	ctx, cancel := context.WithCancel(ctx)
+	ac := &AmplitudeClient{
+		apiKey: apiKey,
+		events: make(chan Event, eventChannelBufferSize),
+		flush:  make(chan chan struct{}),
+		ctx:    ctx,
+		cancel: cancel,
+		client: client,
+	}
+
+	go ac.run()
+
+	return ac
+}
+
+func (a *AmplitudeClient) Publish(evt Event) error {
+	select {
+	case a.events <- evt:
+		return nil
+	default:
+		return fmt.Errorf("event channel buffer of size %d is full", eventChannelBufferSize)
+	}
+}
+
+func (a *AmplitudeClient) Flush() {
+	notify := make(chan struct{})
+	defer close(notify)
+
+	a.flush <- notify
+	// Wait for flush to complete.
+	<-notify
+}
+
+func (a *AmplitudeClient) Close() {
+	a.Flush()
+	a.cancel()
+}
+
+func (a *AmplitudeClient) run() {
+	timer := time.NewTimer(eventSendInterval)
+
+	buffer := make([]Event, eventBufferSize)
+	idx := 0
```

Comment:
seems like this is just len vs cap? why not `make([]Event, 0, eventBufferSize)`?

Reply:
This is a buffer that is allocated once and whose head is overwritten on every flush so that we do not allocate a new buffer on every flush.

Reply:
right, i'm assuming this is just more "copy/paste dfrom amplitude-go" things - in this case, i think it `idx` in `buffer` is serving the same role as `len(buffer)` where the `cap(buffer) == eventBufferSize`